qualtran/_infra/data_types.py

```diff
@@ -809,6 +809,18 @@ def assert_valid_classical_val_array(
         if np.any(val_array >= 2**self.bitsize):
             raise ValueError(f"Too-large classical values encountered in {debug_str}")
 
+    def montgomery_inverse(self, xm, n, p):
+        return xm ** (p - 2) * 2 ** (2 * n) % p
```

Comment:
These values will be huge and O(n) to compute .. prefer the following to keep all intermediate values in $[0, p)$ and the complexity to $\mathcal{O}(\log{p})$ here and below
```suggestion
        return (pow(xm, p - 2, p)) * pow(2, 2 * n, p)) % p
```